### DISCUSSION TITLE
How to get foreign logs (like Nginx) into Seq automatically?

### DISCUSSION THREAD

**Discussion by reddwarf666** [datalust/seq-tickets#1836] [Help] [answered]
I have docker containers for a couple of applications, one of then being Nginx.
How would I get the logs of these applications into Seq?

I have searched and it seems seqcli would be a possibility but before I start going that route I thought to check and see if I missed something.

If seqcli is the only answer, what would be the best way to have that automated?
And will seqcli remember what logs and log entries were already sent so it does not kee sending the same data?

Thanks!

**Comment by liammclennan** [comment, marked as answer]
Hi @reddwarf666,

Anywhere that you have log events in a text file you can [use Fluent Bit](https://blog.datalust.co/tailing-a-log-file-or-folder-with-fluent-bit-and-seq/) to take care of tracking progress and shipping to Seq.

**Reply by reddwarf666** [reply]
Awesome! Thanks Liam, I am going to check that out.

**Reply by liammclennan** [reply]
You're welcome.

**Comment by reddwarf666** [comment]
Hi Liam,
Would you perhaps know the proper setup for Fluent-Bit and Seq? I can have Fluent-Bit read the logs and send to Seq, so reading and communication is fine but the messages appear very weid in Seq, I suspect the format as it is delivered to Seq is not proper. If you have some more knowledge then I'd appreciate that!

**Reply by liammclennan** [reply]
@reddwarf666 this blog post (https://blog.datalust.co/tailing-a-log-file-or-folder-with-fluent-bit-and-seq/) should have everything you need.

**Reply by reddwarf666** [reply]
Thanks! I followed that guid to have a general overview.

The issue is with the Parser or Filter I think

This is my parser:

```
[PARSER]
    Name   nginx
    Format regex
    Regex (?<timestamp>\d{4}\/\d{2}\/\d{2} \d{2}:\d{2}:\d{2}) \[(?<level>.*)\]\ (?<process_id>\d*)#(?<thread_id>\d*): (?<message>.*)
    Time_Key time
    Time_Format %d/%b/%Y:%H:%M:%S %z
```

That regexp is the first one after days of looking and trying that seems to understand the incoming message of nginx and parse the message(s).

This filter is not working it seems, that is where the message "No template provided" seems to originate from:

```
[FILTER]
    Name    modify
    Match   *
    Rename  message @m
```

The message comes in. I just need to parse it right and have filter(s) to transform the fields in the nginx messages.

I was kinda hoping this nginx parsing was already done before with Fluent-Bit but I have yet to fond a regexp that will actually work with nginx messages.

The filter part I need to readup on as I think that transformation is whats ailing me.

**Reply by nblumhardt** [reply]
Hi @reddwarf666, thanks for looping back to us.

In the blog post, the `message` part that appears in:

```
Rename message @m
```

isn't an intrinsic field provided by Fluent Bit - you can see it's a named match from the regex a few paragraphs up in the "parser" section:

```
    Regex       ^(?<time>[^ ]+) (?<message>.+)$
```

Your regex is collecting `process_id` and `thread_id` but there's no `message` match there to rename into `@m`.

What's the ideal portion of the nginx log line to show as the message? If a static message is enough, you can use `Add`:

```
Add @m nginx
```

(I don't believe there's a syntax yet for field values with spaces in them - would need to check up on this.)

Let me know if this helps,
Nick

**Reply by nblumhardt** [reply]
(Relevant docs: https://docs.fluentbit.io/manual/pipeline/filters/modify)

**Reply by reddwarf666** [reply]
Hi Nick,

Yeah, it seems to me I need a better filter. I was reading the Filters page you linked just yesterday evening.

This is the regex I have now for the parser:
```
[PARSER]
    Name   nginx
    Format regex
    Regex (?<timestamp>\d{4}\/\d{2}\/\d{2} \d{2}:\d{2}:\d{2}) \[(?<level>.*)\]\ (?<process_id>\d*)#(?<thread_id>\d*): (?<message>.*)
    Time_Key time
    Time_Format %d/%b/%Y:%H:%M:%S %z
```

Seems to pick up fields

And the filter I have:
```
[FILTER]
    Name parser
    Match nginx
    Key_Name log
    Parser nginx
    
```
  Clearly this is not good enough but I have a feeling I now know what is going on a lot better.
  The thing I need to understand yet is when filtering is happening.
    
  I will muddle along and try some more variants and see if I can get this to work for me!

**Reply by nblumhardt** [reply]
Thanks for the follow up @reddwarf666 - sorry I overlooked that detail of your regex. If  you can post what one of your nginx events looks like in Seq, in the expanded event view with all properties, this might also give us some ideas.

Best regards,
Nick

**Comment by Zombro** [comment]
try these snippets for default error & access logs on fluent-bit 3.0+, openresty / nginx 1.25+


parse
```
[PARSER]
    Name   nginx-access
    Format regex
    Regex ^(?<remote>[^ ]*) (?<host>[^ ]*) (?<user>[^ ]*) \[(?<time>[^\]]*)\] "(?<method>\S+)(?: +(?<path>[^\"]*?)(?: +\S*)?)?" (?<code>[^ ]*) (?<size>[^ ]*)(?: "(?<referer>[^\"]*)" "(?<agent>[^\"]*)")
    Time_Key time
    Time_Format %d/%b/%Y:%H:%M:%S %z

[PARSER]
    Name   nginx-error
    Format regex
    Regex (?<timestamp>\d{4}\/\d{2}\/\d{2} \d{2}:\d{2}:\d{2}) \[(?<level>.*)\]\ (?<process_id>\d*)#(?<thread_id>\d*): (?<message>.*)
    Time_Key time
    Time_Format %d/%b/%Y:%H:%M:%S %z
```


main
```
[INPUT]
    Tag          nginx.access
    Name         tail 
    Parser       nginx-access
    Path         /var/log/nginx/*access*.log
    DB           /var/log/nginx/fluent-bit-nginx-access.db

[INPUT]
    Tag          nginx.error
    Name         tail
    Parser       nginx-error
    Path         /var/log/nginx/error.log
    DB           /var/log/nginx/fluent-bit-nginx-error.db

[FILTER]
    Name Lua
    Match nginx.error
    call seq_nginx_error
    code function seq_nginx_error(tag, timestamp, record) record["@m"] = record.hostname .. ":nginx:" .. record.level .. ": " .. record.message record["@l"] = (record.level == "notice" and "warning" or record.level) return 1, timestamp, record end

[FILTER]
    Name Lua
    Match nginx.access
    call append_seq_header_nginx_access
    code function append_seq_header_nginx_access(tag, timestamp, record) record["@m"] = record.hostname .. ":nginx:access: " .. record.remote .. " " .. record.method .. " " .. record.path .. " " .. record.code return 1, timestamp, record end

[OUTPUT]
    Name             http
    Match            *
    Host             {{ seq.host }}
    Port             {{ seq.port }}
    TLS              {{ seq.tls | ternary("On", "Off") }}
    URI              /api/events/raw?clef
    Header           X-Seq-ApiKey {{ seq.apikey }}
    Format           json_lines
    Json_date_key    @t
    Json_date_format iso8601
    Log_response_payload False
```
